pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/BucketDelayedDeliveryTracker.java

```diff
@@ -0,0 +1,565 @@
+                while (iterator.hasNext()) {
+                    Map.Entry<Long, RoaringBitmap> entry = iterator.next();
+                    byte[] array = new byte[entry.getValue().serializedSizeInBytes()];
+                    entry.getValue().serialize(ByteBuffer.wrap(array));
+                    segmentMetadataBuilder.putDelayedIndexBitMap(entry.getKey(), ByteString.copyFrom(array));
+                    iterator.remove();
+                }
+
+                segmentMetadataList.add(segmentMetadataBuilder.build());
+                segmentMetadataBuilder.clear();
+
+                bucketSnapshotSegments.add(snapshotSegmentBuilder.build());
+                snapshotSegmentBuilder.clear();
+            }
+        }
+
+        SnapshotMetadata bucketSnapshotMetadata = SnapshotMetadata.newBuilder()
+                .addAllMetadataList(segmentMetadataList)
+                .build();
+
+        final int lastSegmentEntryId = segmentMetadataList.size();
+
+        BucketState bucketState = this.createBucket(startLedgerId, endLedgerId);
+        bucketState.setCurrentSegmentEntryId(1);
+        bucketState.setNumberBucketDelayedMessages(numMessages);
+        bucketState.setLastSegmentEntryId(lastSegmentEntryId);
+
+        // Add the first snapshot segment last message to snapshotSegmentLastMessageTable
+        checkArgument(!bucketSnapshotSegments.isEmpty());
+        SnapshotSegment snapshotSegment = bucketSnapshotSegments.get(0);
+        DelayedIndex delayedIndex = snapshotSegment.getIndexes(snapshotSegment.getIndexesCount() - 1);
+        snapshotSegmentLastIndexTable.put(delayedIndex.getLedgerId(), delayedIndex.getEntryId(), bucketState);
+
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Create bucket snapshot, bucketState: {}", dispatcher.getName(), bucketState);
+        }
+
+        CompletableFuture<Long> future = asyncSaveBucketSnapshot(bucketState,
+                bucketSnapshotMetadata, bucketSnapshotSegments);
+        bucketState.setSnapshotCreateFuture(future);
+        future.whenComplete((__, ex) -> {
+            if (ex == null) {
+                bucketState.setSnapshotCreateFuture(null);
+            } else {
+                //TODO Record create snapshot failed
+                log.error("Failed to create snapshot: ", ex);
+            }
+        });
+
+        return future;
+    }
+
+
+    private CompletableFuture<Void> asyncLoadNextBucketSnapshotEntry(BucketState bucketState, boolean isRecover) {
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Load next bucket snapshot data, bucketState: {}", dispatcher.getName(), bucketState);
+        }
+        if (bucketState == null) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        // Wait bucket snapshot create finish
+        CompletableFuture<Long> snapshotCreateFuture =
+                bucketState.getSnapshotCreateFuture().orElseGet(() -> CompletableFuture.completedFuture(-1L));
```

Comment:
This `-1` doesn't really mean anything. I feel that a here is `CompletableFuture<Long>` misleading variable,  it should be declared `CompletableFuture<Void>`